File: backend/app/modules/platform/shared/hierarchy_relation_profile.py
"""Hierarchy relation profiles (task_subtask, document_subdocument, …)."""

from __future__ import annotations

from typing import Any

from app.modules.platform.shared.task_subtask_contract import (
    TASK_SUBTASK_RELATION_KEY,
    TASK_SUBTASK_SEMANTIC_PROFILE,
)

HIERARCHY_PROFILE_SUFFIXES = (
    "_subtask",
    "_subdocument",
    "_subordinate",
    "_subfolder",
    "_subcategory",
)


def _normalize_key(value: Any) -> str:
    return str(value or "").strip()


def is_hierarchy_semantic_profile(profile: str) -> bool:
    normalized = _normalize_key(profile)
    if not normalized:
        return False
    if normalized == TASK_SUBTASK_SEMANTIC_PROFILE:
        return True
    return any(normalized.endswith(suffix) for suffix in HIERARCHY_PROFILE_SUFFIXES)


def is_hierarchy_relation_definition(
    relation: dict[str, Any] | None,
    current_object_type_key: str,
) -> bool:
    current_key = _normalize_key(current_object_type_key)
    if not relation or not current_key:
        return False

    settings = relation.get("settings_json")
    if not isinstance(settings, dict):
        settings = {}

    profile = _normalize_key(settings.get("semantic_profile"))
    relation_key = _normalize_key(relation.get("key"))

    marked_hierarchy = (
        settings.get("is_hierarchy") is True
        or is_hierarchy_semantic_profile(profile)
        or relation_key == TASK_SUBTASK_RELATION_KEY
        or is_hierarchy_semantic_profile(relation_key)
    )
    if not marked_hierarchy:
        return False

    source_key = _normalize_key(relation.get("source_object_type_key"))
    target_key = _normalize_key(relation.get("target_object_type_key"))
    return current_key == source_key or current_key == target_key


def is_hierarchy_subtask_parent_relation_definition(
    relation: dict[str, Any] | None,
    current_object_type_key: str,
) -> bool:
    if not is_hierarchy_relation_definition(relation, current_object_type_key):
        return False

    settings = relation.get("settings_json")
    if not isinstance(settings, dict):
        settings = {}

    parent_side = _normalize_key(settings.get("parent_entity_side") or "source")
    child_side = _normalize_key(settings.get("child_entity_side") or "target")
    source_key = _normalize_key(relation.get("source_object_type_key"))
    target_key = _normalize_key(relation.get("target_object_type_key"))
    current_key = _normalize_key(current_object_type_key)

    if parent_side == "source" and child_side == "target":
        return current_key == source_key
    if parent_side == "target" and child_side == "source":
        return current_key == target_key
    return current_key == source_key


def list_hierarchy_subtask_relation_keys(
    relations: list[dict[str, Any]] | None,
    current_object_type_key: str,
) -> list[str]:
    keys: set[str] = set()
    for relation in relations or []:
        if is_hierarchy_subtask_parent_relation_definition(relation, current_object_type_key):
            key = _normalize_key(relation.get("key"))
            if key:
                keys.add(key)
    return sorted(keys)


def resolve_primary_hierarchy_subtask_relation_key(
    relations: list[dict[str, Any]] | None,
    object_type_key: str,
) -> str:
    keys = list_hierarchy_subtask_relation_keys(relations, object_type_key)
    if not keys:
        return ""
    if TASK_SUBTASK_RELATION_KEY in keys:
        return TASK_SUBTASK_RELATION_KEY
    return keys[0]


def resolve_hierarchy_relation_entity_sides(
    settings_json: dict[str, Any] | None,
) -> tuple[str, str]:
    settings = settings_json if isinstance(settings_json, dict) else {}
    parent_side = _normalize_key(settings.get("parent_entity_side") or "source")
    child_side = _normalize_key(settings.get("child_entity_side") or "target")
    if parent_side == "target" and child_side == "source":
        return "target", "source"
    return "source", "target"


def hierarchy_parent_child_from_edge(
    *,
    source_entity_id,
    target_entity_id,
    parent_side: str,
    child_side: str,
) -> tuple[str, str]:
    source_id = _normalize_key(source_entity_id)
    target_id = _normalize_key(target_entity_id)
    if parent_side == "source" and child_side == "target":
        return source_id, target_id
    return target_id, source_id

File: frontend/src/shared/relation/hierarchyRelationProfile.js
/**
 * Hierarchy relation profiles (task_subtask, document_subdocument, …).
 * Used by Parent Section, card layout filtering, and related-records grouping.
 */

export const TASK_SUBTASK_RELATION_KEY = "task_subtask";
export const TASK_SUBTASK_SEMANTIC_PROFILE = "task_subtask";

const HIERARCHY_PROFILE_SUFFIXES = [
  "_subtask",
  "_subdocument",
  "_subordinate",
  "_subfolder",
  "_subcategory",
];

function normalizeKey(value) {
  return String(value ?? "").trim();
}

/**
 * Self-relation: same object type on source and target sides (by published catalog keys).
 *
 * @param {Record<string, unknown> | null | undefined} relation
 */
export function isSelfRelationDefinition(relation) {
  const sourceKey = normalizeKey(relation?.source_object_type_key);
  const targetKey = normalizeKey(relation?.target_object_type_key);

  return Boolean(sourceKey && targetKey && sourceKey === targetKey);
}

export function isHierarchySemanticProfile(profile) {
  const normalized = normalizeKey(profile);

  if (!normalized) {
    return false;
  }

  if (normalized === TASK_SUBTASK_SEMANTIC_PROFILE) {
    return true;
  }

  return HIERARCHY_PROFILE_SUFFIXES.some((suffix) => normalized.endsWith(suffix));
}

/**
 * Published relation definition marked as hierarchy for the current object type.
 *
 * @param {Record<string, unknown> | null | undefined} relation
 * @param {string | null | undefined} currentObjectTypeKey
 */
export function isHierarchyRelationDefinition(relation, currentObjectTypeKey) {
  const currentKey = normalizeKey(currentObjectTypeKey);

  if (!relation || !currentKey) {
    return false;
  }

  const sourceKey = normalizeKey(relation.source_object_type_key);
  const targetKey = normalizeKey(relation.target_object_type_key);

  if (
    isSelfRelationDefinition(relation) &&
    currentKey === sourceKey &&
    currentKey === targetKey
  ) {
    return true;
  }

  const settings =
    relation.settings_json && typeof relation.settings_json === "object"
      ? relation.settings_json
      : {};

  const profile = normalizeKey(settings.semantic_profile);
  const relationKey = normalizeKey(relation.key);

  const markedHierarchy =
    settings.is_hierarchy === true ||
    isHierarchySemanticProfile(profile) ||
    relationKey === TASK_SUBTASK_RELATION_KEY ||
    isHierarchySemanticProfile(relationKey);

  if (!markedHierarchy) {
    return false;
  }

  return currentKey === sourceKey || currentKey === targetKey;
}

/**
 * Relation eligible for Plan hierarchy picker (self-relation on current type or marked hierarchy).
 *
 * @param {Record<string, unknown> | null | undefined} relation
 * @param {string | null | undefined} currentObjectTypeKey
 */
export function isPlanHierarchyRelationCandidate(relation, currentObjectTypeKey) {
  return isHierarchyRelationDefinition(relation, currentObjectTypeKey);
}

/**
 * Relation field bound to a hierarchy definition (card, table, filters).
 *
 * @param {Record<string, unknown> | null | undefined} field
 * @param {Record<string, unknown> | null | undefined} catalog
 * @param {string | null | undefined} objectTypeKey
 */
export function isHierarchyRelationField(field, catalog, objectTypeKey) {
  if (!field || typeof field !== "object") {
    return false;
  }

  const rawType = String(field.field_type || field.type || "")
    .trim()
    .toLowerCase();

  if (rawType !== "relation") {
    return false;
  }

  const fieldSettings =
    field.settings_json && typeof field.settings_json === "object"
      ? field.settings_json
      : field.settings && typeof field.settings === "object"
        ? field.settings
        : {};

  const relationKey = normalizeKey(fieldSettings.relation_key);

  if (!relationKey) {
    return false;
  }

  const relations = Array.isArray(catalog?.relations) ? catalog.relations : [];
  const catalogRelation = relations.find(
    (item) => normalizeKey(item?.key) === relationKey,
  );

  return Boolean(
    catalogRelation &&
      isHierarchyRelationDefinition(catalogRelation, objectTypeKey),
  );
}

/** @deprecated Use {@link isHierarchyRelationField} — kept for card imports. */
export function isHierarchyRelationFieldForCard(field, catalog, objectTypeKey) {
  return isHierarchyRelationField(field, catalog, objectTypeKey);
}

/** Table projection/columns — same detector as card. */
export function isHierarchyRelationFieldForTable(field, catalog, objectTypeKey) {
  return isHierarchyRelationField(field, catalog, objectTypeKey);
}

/**
 * Current object type acts as parent side of hierarchy (e.g. task as source for subtasks).
 *
 * @param {Record<string, unknown> | null | undefined} relation
 * @param {string | null | undefined} currentObjectTypeKey
 */
export function isHierarchySubtaskParentRelationDefinition(
  relation,
  currentObjectTypeKey,
) {
  if (!isHierarchyRelationDefinition(relation, currentObjectTypeKey)) {
    return false;
  }

  const settings =
    relation.settings_json && typeof relation.settings_json === "object"
      ? relation.settings_json
      : {};

  const parentSide = normalizeKey(settings.parent_entity_side || "source");
  const childSide = normalizeKey(settings.child_entity_side || "target");
  const sourceKey = normalizeKey(relation.source_object_type_key);
  const targetKey = normalizeKey(relation.target_object_type_key);
  const currentKey = normalizeKey(currentObjectTypeKey);

  if (parentSide === "source" && childSide === "target") {
    return currentKey === sourceKey;
  }

  if (parentSide === "target" && childSide === "source") {
    return currentKey === targetKey;
  }

  return currentKey === sourceKey;
}

/**
 * @param {Array<Record<string, unknown>> | null | undefined} relations
 * @param {string | null | undefined} currentObjectTypeKey
 */
export function listHierarchySubtaskRelationKeys(relations, currentObjectTypeKey) {
  const keys = new Set();

  for (const relation of Array.isArray(relations) ? relations : []) {
    if (isHierarchySubtaskParentRelationDefinition(relation, currentObjectTypeKey)) {
      const key = normalizeKey(relation.key);

      if (key) {
        keys.add(key);
      }
    }
  }

  return keys;
}

/**
 * @param {Record<string, unknown> | null | undefined} catalog
 * @param {string | null | undefined} objectTypeKey
 */
export function resolvePrimaryHierarchySubtaskRelationKey(catalog, objectTypeKey) {
  const relations = Array.isArray(catalog?.relations) ? catalog.relations : [];
  const keys = listHierarchySubtaskRelationKeys(relations, objectTypeKey);

  if (!keys.size) {
    return "";
  }

  if (keys.has(TASK_SUBTASK_RELATION_KEY)) {
    return TASK_SUBTASK_RELATION_KEY;
  }

  return [...keys][0];
}

/**
 * @param {Record<string, unknown> | null | undefined} catalog
 * @param {string | null | undefined} objectTypeKey
 */
export function hasHierarchySubtasksFeature(catalog, objectTypeKey) {
  return Boolean(resolvePrimaryHierarchySubtaskRelationKey(catalog, objectTypeKey));
}
